### DISCUSSION TITLE
[Android] Draw models as lines instead of fill - wireframe like

### DISCUSSION THREAD

**Discussion by rrocav3** [google/filament#4558] [Q&A]
Is there a way in filament to show model lines that conform the faces instead of the model faces?

**SceneKit like:**
`node.geometry?.firstMaterial?.fillMode = .lines`
**OpenGL like:**
`glPolygonMode(GL_FRONT_AND_BACK, GL_LINE)`

Expected result:
![image](https://user-images.githubusercontent.com/42677484/131157361-be593846-7b85-43ad-a2d3-171e7e596942.png)

I just want it to be possible in current sceneform maintained, so I opened a thread in their repo:
[sceneform-android-sdk/discussions/147](https://github.com/ThomasGorisse/sceneform-android-sdk/discussions/147)
We saw that there is currently a way to get the bounding box as a wireframe but not the wireframe of polygons that conform the model itself.

Thank you in advance!

**Comment by romainguy** [comment]
OpenGL ES 3.0+ (and Metal/Vk afaik) does not provide `glPolygonMode`. You  must instead use a custom material and do it yourself. An easy solution is to provide barycentric coordinates to your vertex shader and use derivatives to render the wireframe. Here is an article on this topic: https://catlikecoding.com/unity/tutorials/advanced-rendering/flat-and-wireframe-shading/

**Reply by rrocav3** [reply]
The tutorial is a unity reference. Can I export material definition from it and import to filament or I must need to define custom material in filament?

**Reply by romainguy** [reply]
You'll need to create custom materials, but they will be quite similar. Here's another article that uses GLSL so it's closer to Filament's material system: https://tchayen.github.io/posts/wireframes-with-barycentric-coordinates

**Reply by rrocav3** [reply]
So I must try to define it using this reference? https://google.github.io/filament/Materials.html#materialdefinitions
Thank you!

**Reply by jwwalker** [reply]
@romainguy  I don't understand the part about assigning barycentric coordinates as vertex attributes.  Suppose you have 3 triangles like this...

![triangles](https://user-images.githubusercontent.com/730426/131188390-0eb1d3dd-6907-4105-86f6-a6d500a35e71.png)

and you want to assign one of (1, 0, 0), or (0, 1, 0), or (0, 0, 1) to each vertex.  There is no way to do so such that each triangle gets 3 distinct barycentric values for its vertices.

---

After more Googling, I don't think you can use barycentric coordinates without using either a geometry shader or a special extension like GL_NV_fragment_shader_barycentric, neither of which will be possible in Filament.

**Comment by jwwalker** [comment]
An alternate way to do it would be to create a different renderable for the wire frame, with a different index buffer but the same vertex buffer.
